### PR TITLE
CSS3DRenderer: added preserve3d and sortObjects options to prevent flickering

### PR DESCRIPTION
**Pull request by LuYaojing**
此修改版本，兼容IE10及11，禁止preserve3d渲染时，在部分手机上缩放也不会有闪烁的问题。
为CSS3DRenderer对象增加了两个属性：preserve3d，是否启用preserve3d渲染；sortObjects，是否排序对象。启用后近的对象可以遮住远的对象；不启用，会出现远的对象出现在近的对象前面的现象。禁止preserve3d渲染时有效。

### CODE REVIEW

**Comment by dubejf** [comment]
Google translation:

> This modified version, compatible with IE10 and 11, disabled preserve3d rendering, zoom in on part of the phone will not be flicker problem.
>
> CSS3DRenderer target increased to two properties : preserve3d, whether to enable preserve3d rendering ; sortObjects, whether sorting objects. Enabled objects can be hidden away near the object ; it does not, there will be distant object appears in front of the phenomenon of near objects . When disabled preserve3d effective rendering.

**Comment by MasterJames** [comment]
Is it an IE only problem? I have not used or even launched IE in about 5 years. Maybe I'm mistaken and then this I suppose this translation needs rewriting. Still if it works in Chrome it should also be reported to IE team, or be fixed by a pull request only. I vote against cluttering code to make IE bug work arounds. Banishing IE for not being a community player has been my policy for many years. It will take at least a decade of perfection before I feel they've been punished sufficently to no longer fear their tormented ways causing more trouble then its worth.
goto HTML5TEST.com/results/desktop.html
IE 336 vs Chrome 526 it's not worth the effort when they get past Chrome it might be worth it but then there's performance Chrome wins again especially in GL, so let's call this a stackoverflow problem and close it. Thank you.

**Comment by LuYaojing** [comment]
在GL的性能上确实IE表现不好。不过修改这个对象的原因本来是为了解决我的Android4.2上的缩放闪烁问题，结果发现IE也能兼容。
我的英文不好，没法用英文表达，在这里表达歉意。

**Comment by MasterJames** [comment]
Translations are available at Google Translate etc.
"GL performance in IE does not perform well. But the reason to modify the object was supposed to solve the problem of scaling flicker on my Android 4.2 found that IE is also compatible.
My English is not good, can not use the English expression, where expression of regret."

Part of this gets lost in translation? you speak of flicker and scaling here, but sort order in the previous initial note.

Are you clear on what part of threeJS is failing?
Maybe you need to attach an image or something?

你清楚什么threeJS的一部分失败的原因？
也许你需要附加一个图像或东西吗？

**Comment by MasterJames** [comment]
Sorry I didn't see on mobile you've submitted a fix/commit.
MrDoob will have to review it, thanks for your support.

**Comment by LuYaojing** [comment]
screenshot_2015-06-15-13-47-12
screenshot_2015-06-15-13-47-39
这个就是缩放时出现的情况。
This is the case when scaling.
screenshot_2015-06-15-13-47-44
这是正常时的情况。
This is normal situation..

**Comment by yomotsu** [comment]
This might be what LuYaojing means http://slides.com/zz85/cssconf-2014#/22

> Solution: "CSS3DRenderIE" [here](https://github.com/mrdoob/three.js/issues/4783), [here](https://github.com/mrdoob/three.js/pull/3161/files) and [here](https://github.com/Instrument/google-io-2014-experiment/blob/master/static/js/vendor/CSS3DRendererIE.js)

**Comment by MasterJames** [comment]
Yes okay I'm understanding this this better now, thanks.
The sorting is wrong for CSS3DRender in this link but it's an older version mind you?
http://mrdoob.com/lab/javascript/threejs/css3d/
THREE.CSS3DRenderer 51dev
Layering order sorting in the latest Chrome too.
I think there are better times for preferring one technique over the alternatives.

Are you saying this 'commit' has solved the problem of you?
你是说这个'提交'已经解决了你的问题吗？

**Comment by LuYaojing** [comment]
在我测试的设备上，问题得到解决。
On the equipment I tested, the problem is solved..

**Comment by MasterJames** [comment]
Meaning it works with these 'commit' changes only, or without them.
As in close this ticket or you should merge it first?
这意味着它可以与这些“承诺”只更改，或没有他们。
在靠近这张票还是应该先合并呢？

**Comment by LuYaojing** [comment]
我觉得应该能够满足更多的应用场景。Threejs有一些模块是不需要WebGL的，因此只能满足一些简单的应用场景，可那依然非常有价值，我们并不总是工作在复杂的场景中。所以我觉得应该加入这些功能。
I think it should be able to meet more application scenarios. Threejs some modules do not need WebGL, so can only meet some simple application scenarios, but it is still very valuable, we do not always work in complex scenes. So I think I should add these features to me.

**Comment by zz85** [comment]
TLDR: submitter added `preserve3d` and `sortObjects` options to prevent flickering.

I'm not sure to :laughing: or :cry: seeing things lost in translation (unintentional pun!)...

**Comment by mrdoob** [comment]
Maybe #3161 was a better solution?

**Comment by LuYaojing** [comment]
在这里说一个无关的事情。
发现一个网站，该网站的代码与Threejs的相似度很高，但是，却声称是自己独立开发的。这个是不是已经违反了开源授权协议。
这是网址：www.mms3d.cn(http://118.123.8.115:6703/webgl/js/mms3D.R70.js)。
该文件非法复制于Threejs。

An unrelated things said here.
Found a site that the site's code and Threejs a high degree of similarity, but it claims to be its own independent development. This is already a violation of open source licensing.
This is the website: www.mms3d.CN (http://118.123.8.115:6703/webgl/js/mms3D.R70.js).
it is a torts and it copy from the sources of threejs。

**Comment by mrdoob** [comment]
Hehe! That's a first. They are stuck with r70 anyway 😀

File: examples/js/renderers/CSS3DRenderer.js [hunk 1]
```diff
@@ -36,41 +41,134 @@ THREE.CSS3DSprite.prototype.constructor = THREE.CSS3DSprite;
...
+			cache.preserve3d = this._preserve3d = value;
+			
+			if ( this._supportPreserve3d === undefined ){
+			
+				var testDiv = document.createElement( 'div' ).style;
+			
+				if ( testDiv.WebkitTransformStyle !== undefined ) testDiv.WebkitTransformStyle = 'preserve-3d';
+				if ( testDiv.MozTransformStyle !== undefined ) testDiv.MozTransformStyle = 'preserve-3d';
+				if ( testDiv.oTransformStyle !== undefined ) testDiv.oTransformStyle = 'preserve-3d';
+				if ( testDiv.transformStyle !== undefined )testDiv.transformStyle = 'preserve-3d';
```
Comment (vaseker) [inline examples/js/renderers/CSS3DRenderer.js]: Forgot \s after `)`

if ( testDiv.transformStyle !== undefined __)testDiv__

**Comment by Mugen87** [comment]
Should solved by #10988
